src/librustc_trans/back/link.rs

```diff
@@ -823,7 +827,8 @@ fn link_args(cmd: &mut Linker,
 
     // If we're building a dynamic library then some platforms need to make sure
     // that all symbols are exported correctly from the dynamic library.
-    if crate_type != config::CrateTypeExecutable {
+    if crate_type != config::CrateTypeExecutable ||
+       sess.target.target.options.is_like_emscripten {
```

Comment:
I don't know... In case with emscripten, definitions of "dynamic library" / "executable" are less precise, so if we export some functions to JS, it can still count as "dynamic library" I guess?